Make FreeType select a size for fixed sized fonts

Green-Sky opened the pull request:
### This PR implements:
- select size for fixed sized unscalable bitmap fonts (sbix and CBDT/CBLC)
- rescale bitmap before adding it to atlas

This enables imgui+freetype to just use the system provided emojie fonts. (NotoColorEmoji and AppleColorEmoji)

### TODO:
- [x] Better rescaling using [box sampling](https://en.wikipedia.org/wiki/Image_scaling#Box_sampling)
- [ ] More testing
  - [ ] niche api usages
  - [x] more fonts
  - [x] multi bitmap fonts without fallbacks (apple emoji)
  - [x] em square pr (hm looks the same, but that pr has uncommited changes)
- [ ] figure out why emojies are generally offset upwards

---

This should make CBDT fonts like NotoColorEmoji and AppleColorEmoji play nice. Bitmap fonts are often preferred by distributors over COLRv1, because they are more universally supported, while also being smaller than SVGinOT fonts.

see https://github.com/ocornut/imgui/issues/6302

---

### built-in font + system provided NotoColorEmoji (109px) at font size 13:
<img width="844" height="783" alt="image" src="https://github.com/user-attachments/assets/a03285f9-a358-4ec2-8d4e-c85b3e0c7be7" />

### built-in font + system provided NotoColorEmoji (109px) at font size 26:
<img width="1165" height="836" alt="image" src="https://github.com/user-attachments/assets/2308b2d4-ce8b-4dc6-a95c-dad16c28b9ab" />


<details><summary>Old images</summary>

### built-in font + system provided NotoColorEmoji (109px) at font size 13:
<img width="953" height="775" alt="image" src="https://github.com/user-attachments/assets/3f56aa49-e464-4100-8df5-8bcf6d5c63f1" />

### built-in font + system provided NotoColorEmoji (109px) at font size 26:
<img width="1158" height="857" alt="image" src="https://github.com/user-attachments/assets/49cd1821-a0db-471a-ac8b-b61bbbe5cd7f" />

</details>

<details><summary>Old Old images</summary>

### built-in font + system provided NotoColorEmoji (109px) at font size 13:
<img width="878" height="878" alt="image" src="https://github.com/user-attachments/assets/da8ab2c7-4738-47ac-b21b-e2fa4a171ff8" />

### built-in font + system provided NotoColorEmoji (109px) at font size 26:
<img width="1014" height="886" alt="image" src="https://github.com/user-attachments/assets/59b89021-e007-4ef4-a2b2-17fe96cf0e16" />

</details>

ocornut commented:
I know this is unfinished but please make sure you follow the local coding style (brace placement, c-style cast, uppercase first letter in comment, etc.) as this really facilitate me looking at PRs. Thank you.

Green-Sky commented:
I have now also added downscaling using Point Sampling / Nearest Neighbor.
Formatting is probably correct too.

Any TODOs left in the code now need to be resolved, before this pr is ready. Feel free to address them.

The biggest TODO is https://en.wikipedia.org/wiki/Image_scaling#Box_sampling .
I wanted to have NN first, so I can compare and get the sizing/placement of the texture correct first, which it now hopefully is.

The placement sometimes seems odd, especially mixing fonts, but I guess that relates to #8857 , which I have not yet reviewed/tested this code with. There will probably be light conflicts.

edit: I also updated the images in the OP.

Green-Sky commented:
Ceiling bitmap size, instead of Flooring it.

| size | floor | ceil |
|--------|--------|--------|
| 13 | <img width="436" height="120" alt="image" src="https://github.com/user-attachments/assets/e5f7c433-1840-4480-abd7-ad72ea83cabd" /> | <img width="438" height="124" alt="image" src="https://github.com/user-attachments/assets/2aaa5705-366d-4d88-a861-13e7c432e4f5" /> |
| 26 | <img width="422" height="104" alt="image" src="https://github.com/user-attachments/assets/46c00131-54e7-4640-b6f4-be640db356a8" /> | <img width="417" height="111" alt="image" src="https://github.com/user-attachments/assets/0a9dddae-c275-4d53-b2dd-05323f184dc8" /> | 

More in line with the rest of the code. Also feels placed more correctly. NN scaling looking decimating in all scenarios, naturally.

Green-Sky commented:
I am contemplating using the downscale function to also just upscale them before adding them to the atlas.

Making use of the current size selection bug, selecting the next smaller size:

### upscaled using Point sampling

<img width="354" height="244" alt="image" src="https://github.com/user-attachments/assets/e235123b-a621-45ad-89a0-1c79d37f3bf2" />

### upscaled using Box sampling

<img width="358" height="241" alt="image" src="https://github.com/user-attachments/assets/1f6b73bc-6463-4dca-b710-d022c20b9f31" />

Box sampling might be good enough.

Green-Sky commented:
We are treating the memory allocated by `ImVector` as 4byte aligned. How sure can we be this is always true?

edit: `__STDCPP_DEFAULT_NEW_ALIGNMENT__` defaults to `sizeof(size_t)`, which is "The bit width of std::size_t is not less than 16", so 2bytes.
So on some obscure platforms imgui might not work.
